src/net/netsyscall.c

```diff
@@ -423,8 +436,13 @@ static sysreturn sock_read_bh_internal(netsock s, thread t, void * dest,
                 length -= xfer;
                 xfer_total += xfer;
                 dest = (char *) dest + xfer;
-                if ((s->sock.type == SOCK_STREAM) && !(flags & MSG_PEEK))
+                if ((s->sock.type == SOCK_STREAM) && !(flags & MSG_PEEK)) {
+                    if (!blocked)
+                        lwip_lock();
```

Comment:
This could deadlock with a parallel thread where sock_read_bh_internal() is called with lwIP already locked and is trying to acquire s->consume_lock. To prevent this deadlock, the lwIP lock should be acquired before s->consume_lock.